### DISCUSSION TITLE
FAQ: What's the relationship between SeedTool, KeyTool and HD-wallet related BIPs?

### DISCUSSION THREAD

**Discussion by ChristopherA** [BlockchainCommons/Gordian-Developer-Community#17] [General & Announcements]
## SeedTool: Entropy and Master Seeds

* https://github.com/blockchainCommons/bc-seedtool-cli (available now, currently in late alpha but should be feature complete beta this week or next.)

SeedTool  is focused on taking raw entropy and reliably and safely turning it into a specific form we call a Master Seed (sometimes itself called entropy because it too is random, but we believe this is the wrong name for it). Master Seeds need to be able to be secured offline, which is is why it supports export to [BIP-39](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki) and to the emerging [SSKR](https://github.com/BlockchainCommons/Research/blob/master/papers/bcr-2020-011-sskr.md) (Shamir Secret Key Recovery) standards, and via [UR](https://github.com/BlockchainCommons/Research/blob/master/papers/bcr-2020-005-ur.md) self-describing formats. Master Seeds need to be in a form that can be used as input by various Hierarchical Deterministic (HD) schemes for creating child keys.

SeedTool leverages a number of open source libraries for all wallets that create or use Master Seeds, thus also is intended as a reference tool to exercise the features of those libraries.

## KeyTool & Bitcoin BIPS 

* https://github.com/blockchainCommons/bc-keytool-cli (coming soon, currently in private alpha)

KeyTool takes Master Seeds (sometimes called entropy by libraries like LibWally) and converts them into a Master Key (a private key plus a chain code secret), and then derives from that child trees of private keys. For bitcoin's secp256k1 that is effectively [BIP-32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki). But there are other standards for specific ways of defining that tree, in particular legacy bitcoin-core, electrum wallets, and other popular legacy wallets, the [BIP-44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki), [BIP-32](https://github.com/bitcoin/bips/blob/master/bip-0045.mediawiki), [BIP-32](https://github.com/bitcoin/bips/blob/master/bip-0049.mediawiki), and [BIP-84](https://github.com/bitcoin/bips/blob/master/bip-0084.mediawiki), and an unwritten [BIP48](https://github.com/BlockchainCommons/AirgappedSigning/discussions/16) for how keys work with hardware keys. Keytool handles those "conventions" as well, that include cointype, sub-account, etc. metdata. It also supports export of child keys in a variety of formats, in particular various *prv/*pub text formats and [UR](https://github.com/BlockchainCommons/Research/blob/master/papers/bcr-2020-005-ur.md) self-describing formats. 

Like SeedTool this app is a reference app to allow testing of key derivation libraries.

## KeyTool & Other Standards

Other blockchains use Master Seeds as well, as evidenced by the fact that Ledger & Trezor demonstrate support of securing keys for them. But it isn't clear if they create the Master Seed the same way, or if they use a different technique. And we know that they can't use the same [BIP-32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki) math while using 25519 curves to create child keys, and that there are at least two specs for how to do this, but I'm not sure which are in production. Finally, we don't know the "conventions" within the 25519 blockchain wallet community for cointype, sub-account and other metadata.

Trezor & Ledger are can also use Master Seeds to create GPG and SSH keys, we should document how they do that. I also would like us to be able to derive keys for use with Tor hidden onion servers and v3 client certificates.

Anywhere a current hardware wallet can take a Master Seed and use it to create keys for a cryptographic protocol, KeyTool should eventually be able to do it.

/cc: @wolfmcnally - can you comment on this for technical accuracy
/cc: @shannona - after some review this is a candidate for a FAQ answer in this repository

**Comment by wolfmcnally** [comment]
LGTM
